src/main/scala/core/reservations/ReservationService.scala

```diff
@@ -0,0 +1,132 @@
+  def getReservations: Future[Seq[Reservation]] = reservationStorage.getReservations
+
+  def getReservation(id: Long): Future[Option[Reservation]] = reservationStorage.getReservation(id)
+
+  def makeReservation(reservationForm: ReservationForm): Future[(Option[ReservationSummary], Option[String])] = {
+    val screeningData = Await.result(screeningStorage.getScreeningDetails(reservationForm.screeningID), Duration.Inf)
+    if (screeningData.isEmpty) return Future((None, Some("Cannot find requested screening")))
+
+    val reservationTime = new Timestamp(Calendar.getInstance().getTimeInMillis)
+    if (reservationForm.seats.isEmpty)
+      return Future((None, Some("Reservation must have at least 1 seat")))
+
+    if (!checkReservationTime(reservationTime, screeningData.get._1.screeningTime))
+      return Future((None, Some("There are less than 15 minutes to screening")))
+
+    if (!checkFullName(reservationForm.name, reservationForm.surname))
+      return Future((None, Some("Illegal name or surname")))
+
+    val paymentTime = new Timestamp(
+      if ((reservationTime.getTime + 3 * 24 * 60 * 60 * 1000) < screeningData.get._1.screeningTime.getTime)
+        reservationTime.getTime + 5 * 24 * 60 * 60 * 1000
+      else screeningData.get._1.screeningTime.getTime
+    )
+    val seatsWithoutID = reservationForm.seats.map(s => Seat(None, 0, s.row, s.index, ticketTypeMapper(s.ticketType)))
+    val takenSeats = Await.result(seatStorage.takenSeats(reservationForm.screeningID), Duration.Inf)
+
+    if (!checkRowsValues(seatsWithoutID, takenSeats, screeningData.get._3))
+      return Future((None, Some("There is either single seat space in row or seats indices are out of bounds")))
+
+    val reservation = Await.result(
+      reservationStorage.saveReservation(
+        Reservation(
+          None,
+          reservationForm.screeningID,
+          reservationForm.name,
+          reservationForm.surname,
+          reservationTime,
+          "UNPAID"
+        )
+      ),
+      Duration.Inf
+    )
+
+    val seatsToReserve = seatsWithoutID.map(s => Seat(None, reservation.id.get, s.row, s.index, s.price))
+
+    seatStorage.saveSeats(seatsToReserve)
+
+    val totalPrice = seatsWithoutID.foldLeft(0.0)(_ + _.price)
+
+    Future((Some(ReservationSummary(reservation.id.get, totalPrice, paymentTime)), None))
+  }
+
+  def ticketTypeMapper(ticketType: String): Double =
+    ticketType match {
+      case "adult"   => 25.0
+      case "student" => 18.0
+      case "child"   => 12.5
+      case _         => -1.0
+    }
+
+  def checkReservationTime(reservationTime: Timestamp, screeningTime: Timestamp): Boolean = {
+    (reservationTime.getTime + (15 * 60 * 10)) < screeningTime.getTime
+  }
+
+  def checkFullName(name: String, surname: String): Boolean = {
+    name.matches("""^\p{Lu}[\p{L}&&[^\p{Lu}]]{2,}""") &&
+    surname.matches("""^\p{Lu}[\p{L}&&[^\p{Lu}]]{2,}(-\p{Lu}[\p{L}&&[^\p{Lu}]]{2,})?""")
+  }
```

Comment:
Compiling regex is expensive and I should have constructed it once. It would be better to declare both method and variable in object (e.g in ReservationValidationService).